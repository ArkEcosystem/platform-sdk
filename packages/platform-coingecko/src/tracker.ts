import { Contracts, Exceptions, Utils } from "@arkecosystem/platform-sdk";

import { HistoricalPriceTransformer } from "./transformers/historical-price-transformer";
import { HistoricalVolumeTransformer } from "./transformers/historical-volume-transformer";
import { MarketTransformer } from "./transformers/market-transformer";

export class PriceTracker implements Contracts.PriceTracker {
	private readonly baseUrl: string = "https://api.coingecko.com/api/v3";

	private readonly tokenLookup: Contracts.KeyValuePair = {};

	public async verifyToken(token: string): Promise<boolean> {
		const tokenId = await this.getTokenId(token);

		try {
			const body = await Utils.getJSON(`${this.baseUrl}/simple/price`, {
				ids: tokenId,
				vs_currencies: "BTC",
			});

			return !!body[tokenId];
		} catch {
			return false;
		}
	}

	public async marketData(token: string): Promise<Contracts.MarketDataCollection> {
		const tokenId = await this.getTokenId(token);

		const body = await Utils.getJSON(`${this.baseUrl}/coins/${tokenId}`);

		return new MarketTransformer(body.market_data).transform({});
	}

	public async historicalPrice(options: Contracts.HistoricalPriceOptions): Promise<Contracts.HistoricalData> {
		const tokenId = await this.getTokenId(options.token);

		const body = await Utils.getJSON(`${this.baseUrl}/coins/${tokenId}/market_chart`, {
			vs_currency: options.currency,
			days: options.days,
		});

		return new HistoricalPriceTransformer(body).transform(options);
	}

	public async historicalVolume(options: Contracts.HistoricalVolumeOptions): Promise<Contracts.HistoricalData> {
		const tokenId = await this.getTokenId(options.token);

		const body = await Utils.getJSON(`${this.baseUrl}/coins/${tokenId}/market_chart/range`, {
			id: options.token,
			vs_currency: options.currency,
			from: Utils.dayjs().subtract(options.days, "d").unix(),
			to: Utils.dayjs().unix(),
		});

		return new HistoricalVolumeTransformer(body).transform(options);
	}

	public async dailyAverage(options: Contracts.DailyAverageOptions): Promise<number> {
		throw new Exceptions.NotImplemented(this.constructor.name, "dailyAverage");
	}

	private async getTokenId(token): Promise<string> {
		if (Object.keys(this.tokenLookup).length > 0) {
			return this.tokenLookup[token.toUpperCase()];
		}

		const uri = `${this.baseUrl}/coins/list`;
		const body = await Utils.getJSON(uri);

		for (const value of Object.values(body)) {
			// @ts-ignore
			this.tokenLookup[value.symbol.toUpperCase()] = value.id;
		}

		return this.tokenLookup[token.toUpperCase()];
	}
}
