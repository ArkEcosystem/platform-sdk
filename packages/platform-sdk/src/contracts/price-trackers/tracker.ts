import { DailyAverageOptions, HistoricalData, HistoricalPriceOptions, HistoricalVolumeOptions } from "./historical";
import { MarketDataCollection } from "./market";

/**
 *
 *
 * @export
 * @interface PriceTracker
 */
export interface PriceTracker {
	/**
	 *
	 *
	 * @param {string} token
	 * @returns {Promise<boolean>}
	 * @memberof PriceTracker
	 */
	verifyToken(token: string): Promise<boolean>;

	/**
	 *
	 *
	 * @param {string} token
	 * @returns {Promise<MarketDataCollection>}
	 * @memberof PriceTracker
	 */
	marketData(token: string): Promise<MarketDataCollection>;

	/**
	 *
	 *
	 * @param {HistoricalPriceOptions} options
	 * @returns {Promise<HistoricalData>}
	 * @memberof PriceTracker
	 */
	historicalPrice(options: HistoricalPriceOptions): Promise<HistoricalData>;

	/**
	 *
	 *
	 * @param {HistoricalVolumeOptions} options
	 * @returns {Promise<HistoricalData>}
	 * @memberof PriceTracker
	 */
	historicalVolume(options: HistoricalVolumeOptions): Promise<HistoricalData>;

	/**
	 *
	 *
	 * @param {DailyAverageOptions} options
	 * @returns {Promise<number>}
	 * @memberof PriceTracker
	 */
	dailyAverage(options: DailyAverageOptions): Promise<number>;
}
