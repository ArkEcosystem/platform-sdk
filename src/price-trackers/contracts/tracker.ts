import { DailyAverageOptions, HistoricalData, HistoricalPriceOptions, HistoricalVolumeOptions } from "./historical";
import { MarketDataCollection } from "./market";

export interface PriceTracker {
	verifyToken(token: string): Promise<boolean>;

	marketData(token: string): Promise<MarketDataCollection>;

	historicalPrice(options: HistoricalPriceOptions): Promise<HistoricalData>;

	historicalVolume(options: HistoricalVolumeOptions): Promise<HistoricalData>;

	dailyAverage(options: DailyAverageOptions): Promise<number>;
}
