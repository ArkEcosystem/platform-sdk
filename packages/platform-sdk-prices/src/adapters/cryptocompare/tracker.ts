import { Utils } from "@arkecosystem/platform-sdk";

import { CURRENCIES } from "../../config";
import {
	DailyAverageOptions,
	HistoricalData,
	HistoricalPriceOptions,
	HistoricalVolumeOptions,
} from "../../contracts/historical";
import { MarketDataCollection } from "../../contracts/market";
import { PriceTracker } from "../../contracts/tracker";
import { HistoricalPriceTransformer } from "./transformers/historical-price-transformer";
import { HistoricalVolumeTransformer } from "./transformers/historical-volume-transformer";
import { MarketTransformer } from "./transformers/market-transformer";

export class CryptoCompare implements PriceTracker {
	private readonly baseUrl: string = "https://min-api.cryptocompare.com";

	public async verifyToken(token: string): Promise<boolean> {
		try {
			const uri = `${this.baseUrl}/data/price`;
			const body = await Utils.getJSON(uri, {
				fsym: token,
				tsyms: "BTC",
			});

			return !!body.BTC;
		} catch {
			return false;
		}
	}

	public async marketData(token: string): Promise<MarketDataCollection> {
		const uri = `${this.baseUrl}/data/pricemultifull`;
		const body = await Utils.getJSON(uri, {
			fsyms: token,
			tsyms: Object.keys(CURRENCIES).join(","),
		});

		return new MarketTransformer(body.RAW && body.RAW[token] ? body.RAW[token] : {}).transform({});
	}

	public async historicalPrice(options: HistoricalPriceOptions): Promise<HistoricalData> {
		const body = await Utils.getJSON(`${this.baseUrl}/data/histo${options.type}`, {
			fsym: options.token,
			tsym: options.currency,
			toTs: Math.round(new Date().getTime() / 1000),
			limit: options.days,
		});

		return new HistoricalPriceTransformer(body.Data).transform(options);
	}

	public async historicalVolume(options: HistoricalVolumeOptions): Promise<HistoricalData> {
		const body = await Utils.getJSON(`${this.baseUrl}/data/histo${options.type}`, {
			fsym: options.token,
			tsym: options.currency,
			toTs: Math.round(new Date().getTime() / 1000),
			limit: options.days,
		});

		return new HistoricalVolumeTransformer(body.Data).transform(options);
	}

	public async dailyAverage(options: DailyAverageOptions): Promise<number> {
		const response = await Utils.getJSON(`${this.baseUrl}/data/dayAvg`, {
			fsym: options.token,
			tsym: options.currency,
			toTs: Utils.dayjs(options.timestamp).unix(),
		});

		return response[options.currency.toUpperCase()];
	}
}
