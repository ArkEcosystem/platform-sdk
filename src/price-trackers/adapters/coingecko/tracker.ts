import { NotImplemented } from "../../../exceptions";
import { KeyValuePair } from "../../../types";
import { dayjs } from "../../../utils/dayjs";
import { getJSON } from "../../../utils/get-json";
import {
	DailyAverageOptions,
	HistoricalData,
	HistoricalPriceOptions,
	HistoricalVolumeOptions,
} from "../../contracts/historical";
import { MarketDataCollection } from "../../contracts/market";
import { PriceTracker } from "../../contracts/tracker";
import { HistoricalPriceTransformer } from "./transformers/historical-price-transformer";
import { HistoricalVolumeTransformer } from "./transformers/historical-volume-transformer";
import { MarketTransformer } from "./transformers/market-transformer";

export class CoinGecko implements PriceTracker {
	private readonly baseUrl: string = "https://api.coingecko.com/api/v3";

	private readonly tokenLookup: KeyValuePair = {};

	public async verifyToken(token: string): Promise<boolean> {
		const tokenId = await this.getTokenId(token);

		try {
			const body = await getJSON(`${this.baseUrl}/simple/price`, {
				ids: tokenId,
				vs_currencies: "BTC",
			});

			return !!body[tokenId];
		} catch {
			return false;
		}
	}

	public async marketData(token: string): Promise<MarketDataCollection> {
		const tokenId = await this.getTokenId(token);

		const body = await getJSON(`${this.baseUrl}/coins/${tokenId}`);

		return new MarketTransformer(body.market_data).transform({});
	}

	public async historicalPrice(options: HistoricalPriceOptions): Promise<HistoricalData> {
		const tokenId = await this.getTokenId(options.token);

		const body = await getJSON(`${this.baseUrl}/coins/${tokenId}/market_chart`, {
			vs_currency: options.currency,
			days: options.days,
		});

		return new HistoricalPriceTransformer(body).transform(options);
	}

	public async historicalVolume(options: HistoricalVolumeOptions): Promise<HistoricalData> {
		const tokenId = await this.getTokenId(options.token);

		const body = await getJSON(`${this.baseUrl}/coins/${tokenId}/market_chart/range`, {
			id: options.token,
			vs_currency: options.currency,
			from: dayjs().subtract(options.days, "d").unix(),
			to: dayjs().unix(),
		});

		return new HistoricalVolumeTransformer(body).transform(options);
	}

	public async dailyAverage(options: DailyAverageOptions): Promise<number> {
		throw new NotImplemented(this.constructor.name, "dailyAverage");
	}

	private async getTokenId(token): Promise<string> {
		if (Object.keys(this.tokenLookup).length > 0) {
			return this.tokenLookup[token.toUpperCase()];
		}

		const uri = `${this.baseUrl}/coins/list`;
		const body = await getJSON(uri);

		for (const value of Object.values(body)) {
			// @ts-ignore
			this.tokenLookup[value.symbol.toUpperCase()] = value.id;
		}

		return this.tokenLookup[token.toUpperCase()];
	}
}
