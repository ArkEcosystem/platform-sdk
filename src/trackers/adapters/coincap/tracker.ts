import { HistoricalOptions, Tracker } from "../../contracts";
import { HistoricalTransformer } from "./historical-transformer";
import { MarketTransformer } from "./market-transformer";

export class CoinCap implements Tracker {
	private readonly baseUrl: string = "...";

	public async verifyToken(token: string): Promise<boolean> {
		return true;
	}

	public async getMarketData(token: string): Promise<object> {
		return new MarketTransformer().transform({});
	}

	public async getHistoricalData(
		token: string,
		currency: string,
		opts: HistoricalOptions = { type: "day", dateFormat: "DD.MM" },
	): Promise<object> {
		return new HistoricalTransformer().transform({});
	}
}
