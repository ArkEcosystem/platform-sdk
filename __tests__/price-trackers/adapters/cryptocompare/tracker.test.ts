import { CryptoCompare } from "../../../../src/price-trackers/adapters/cryptocompare/tracker";

let subject: CryptoCompare;

beforeEach(() => (subject = new CryptoCompare()));

describe("CryptoCompare", function () {
	it("should get the daily average", async () => {
		await expect(
			subject.dailyAverage({ token: "ark", currency: "btc", timestamp: Date.now().valueOf() }),
		).resolves.toMatchSnapshot();
	});
});
