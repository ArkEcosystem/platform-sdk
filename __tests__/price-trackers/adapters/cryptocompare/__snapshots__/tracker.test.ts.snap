// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoCompare should get the daily average 1`] = `0.00002336`;
